pkg/crdupdater/configmap.go

```diff
@@ -29,21 +29,149 @@ import (
 	"github.com/juju/errors"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
 )
 
+// ConfigMap stores the configuration to add or delete the config map object
+type ConfigMap struct {
+	config        *CommonConfig
+	deployer      *util.DeployerHelper
+	configMaps    []*components.ConfigMap
+	oldConfigMaps map[string]*corev1.ConfigMap
+	newConfigMaps map[string]*corev1.ConfigMap
+}
+
+// NewConfigMap returns the config map
+func NewConfigMap(config *CommonConfig, configMaps []*components.ConfigMap) (*ConfigMap, error) {
+	deployer, err := util.NewDeployer(config.kubeConfig)
+	if err != nil {
+		return nil, errors.Annotatef(err, "unable to get deployer object for %s", config.namespace)
+	}
+	return &ConfigMap{
+		config:        config,
+		deployer:      deployer,
+		configMaps:    configMaps,
+		oldConfigMaps: make(map[string]*corev1.ConfigMap, 0),
+		newConfigMaps: make(map[string]*corev1.ConfigMap, 0),
+	}, nil
+}
+
+// buildNewAndOldObject builds the old and new config map
+func (c *ConfigMap) buildNewAndOldObject() error {
+	// build old config map
+	oldConfigMaps, err := c.list()
+	if err != nil {
+		return errors.Annotatef(err, "unable to get config maps for %s", c.config.namespace)
+	}
+	for _, oldConfigMap := range oldConfigMaps.(*corev1.ConfigMapList).Items {
+		c.oldConfigMaps[oldConfigMap.GetName()] = &oldConfigMap
+	}
+
+	// build new config map
+	for _, newConfigMap := range c.configMaps {
+		newConfigMapKube, err := newConfigMap.ToKube()
+		if err != nil {
+			return errors.Annotatef(err, "unable to convert config map %s to kube %s", newConfigMap.GetName(), c.config.namespace)
+		}
+		c.newConfigMaps[newConfigMap.GetName()] = newConfigMapKube.(*corev1.ConfigMap)
+	}
+
+	return nil
+}
+
+// add adds the config map
+func (c *ConfigMap) add(isPatched bool) (bool, error) {
```

Comment:
It will be used only in replication controller and deployment type where if there is any configmap or secret updated, it will automatically patch the rc and deployment.

Reply:
From external point of view, updater will call buildneworoldobject, add and remove methods.. buildneworoldobject will internally call list or get object, the add method will internally call patch method if it is already exist and remove method will internally call delete method, if it not exist in new object. That's why in add method, we are passing the isPatched..